src/Server/TCPHandler.cpp

```diff
@@ -886,13 +886,21 @@ AsynchronousInsertQueue::PushResult TCPHandler::processAsyncInsertQuery(Asynchro
     using PushResult = AsynchronousInsertQueue::PushResult;
 
     startInsertQuery();
-    SquashingTransform squashing(0, query_context->getSettingsRef().async_insert_max_data_size);
+    Squashing squashing(state.input_header, 0, query_context->getSettingsRef().async_insert_max_data_size);
 
     while (readDataNext())
     {
-        auto result = squashing.add(std::move(state.block_for_insert));
-        if (result)
+        squashing.header = state.block_for_insert;
+        auto planned_chunk = squashing.add({state.block_for_insert.getColumns(), state.block_for_insert.rows()});
+        if (planned_chunk.hasChunkInfo())
         {
+            Chunk result_chunk = squashing.squash(std::move(planned_chunk));
+            ColumnsWithTypeAndName cols;
+            if (result_chunk.hasColumns() && state.block_for_insert)
```

Comment:
That code repeats several times. As I understand we construct a `Block` from `Chunk` according to the types annotations from some header.
May be some function could help here?